### ISSUE TITLE
general formats of character and pattern representation

### ISSUE THREAD

**Issue by LinguList** [pylogeny/pylogeny#7]
with [parsimony](https://github.com/pylogeny) and [likelihood](https://github.com/pylogeny/likelihood) we have now working examples for the basic algorithms (plus a very nice site concordance algorithm) for character evolution.

**Comment by LinguList** [comment]
To keep our models GENERAL, we need to assume the following structure for patterns in all cases:

```python
patterns = {
    "key1": {
        "taxon1": ["a"],
        "taxon2": ["Ø"],
        "taxon3": ["a"],
        "taxon4": ["a"],
   },
    "key2": {
        "taxon1": ["a"],
        "taxon2": ["a"],
        "taxon3": ["a", "b"],
        "taxon4": ["Ø"],
   },
}
```

Note that for a given key, we always need data for all languages in our pattern, and if data is missing, we use a missing character state value `Ø` (as a default).

**Comment by LinguList** [comment]
Multiple character states for one taxon are handled by expanding the list.

**Comment by LinguList** [comment]
This is crucial, as both for parsimony AND for likelihood, we want to be able to handle synonyms, and mathematically, this is not a problem, since it is all about probabilities (in parsimony, it is a problem, but this means the best fitting character will be used, the other will be ignored). Likelihood is anyway a better model for evolution here.

**Comment by LinguList** [comment]
Maybe, @PhyloStar and @xrotwang, it is useful to make a Python class for characters, to be able to check this behaviour is enforced? I would definitely not want to trade of flexibility over simplicity and give up the list (or tuple) representation of multiple synonyms for the same language.

**Comment by xrotwang** [comment]
Why is `patterns` not a `list`? I'd find it very intuitive if a *pattern* is a mapping (i.e. `dict`) from taxa to states. A bunch of patterns should simply be a `list` of these dicts.

**Comment by xrotwang** [comment]
If a pattern needs (or has) metadata like an identifier, it should probably be a class inheriting from `dict`, and adding metadata as attributes?

**Comment by LinguList** [comment]
The reason is that I want to retain information on the cognate set, 
which is the dict's key. If we give this up, we have to rely on order or 
similar, which is not as explicit. So the COGID or the CONCEPT should be 
somehow retrievable for each pattern. In previous analyses, in nexus 
files, this info is always lost, which I find very annoying for 
linguistic purposes.

**Comment by LinguList** [comment]
Yes, in fact, it can have more metadata, or also none, so having metadata as attributes may be useful.

**Comment by xrotwang** [comment]
Yes, I think if your `dict` keys are attributes of the pattern - rather than mnemonics of the calling code to lookup patterns in a convenient way - they should be part of the pattern itself, i.e. `class Pattern(dict)`.
